### PR TITLE
[Material] Add clip property to bottom sheet and theme

### LINKED ISSUE #38827: Add ability to specify Clip behavior for Bottom sheets
## Proposal

Add a clip behavior property to bottom sheet and the bottom sheet theme so that you can enable clipping of the content if necessary.

### PR DESCRIPTION
**Pull request by rami-a**
## Description

Added a property to `BottomSheet` and `BottomSheetThemeData` to customize clip behavior. Also added the argument to the `showBottomSheet` and `showModalBottomSheet` helpers.

Use this property to enable clipping of content when the bottom sheet has a custom shape and the content can extend past this shape. For example, a bottom sheet with rounded corners and an edge-to-edge Image at the top.

## Related Issues

Fixes #38827

## Tests

I added the following tests:

* Updated bottom sheet tests to check that clip is passed through correctly

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/material/scaffold.dart [hunk 1]
```diff
@@ -2427,6 +2434,9 @@ class _StandardBottomSheetState extends State<_StandardBottomSheet> {
         onClosing: widget.onClosing,
         builder: widget.builder,
         backgroundColor: widget.backgroundColor,
+        elevation: widget.elevation,
```
Comment (johnsonmh) [inline packages/flutter/lib/src/material/scaffold.dart]: Hmm were these accidentally left out before? If so, can you add a test that catches the bug? If there is no bug then it's possible this never needed to be passed in the first place and you can leave it out.

**Review by johnsonmh — APPROVED** [review]
LGTM

**Review by willlarche — APPROVED** [review]

**Review by clocksmith — APPROVED** [review]

File: packages/flutter/lib/src/material/scaffold.dart (hunk 1 above)
Comment (rami-a) [reply packages/flutter/lib/src/material/scaffold.dart]: Hmm it actually looks like it's not even possible to hit this code path. Since the `_StandardBottomSheetState` asserts that `animationController` is not null. I think I may simply remove this code path. @HansMuller any idea why this logic was here?

File: packages/flutter/lib/src/material/bottom_sheet.dart [hunk 2]
```diff
@@ -115,6 +116,12 @@ class BottomSheet extends StatefulWidget {
   /// Defaults to null and falls back to [Material]'s default.
   final ShapeBorder shape;
 
+  /// {@macro flutter.widgets.Clip}
+  ///
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/bottom_sheet.dart]: Can you explain here why someone would want to specify something aside from the Clip.none default? The PR's description should cover this as well.

Also; the shape parameter doc notes that it applies to the underlying Material widget. Might say the same thing here.

File: packages/flutter/lib/src/material/scaffold.dart [hunk 3]
```diff
@@ -2398,35 +2404,26 @@ class _StandardBottomSheetState extends State<_StandardBottomSheet> {
...
-          BottomSheet(
-            animationController: widget.animationController,
-            enableDrag: widget.enableDrag,
-            onClosing: widget.onClosing,
-            builder: widget.builder,
-            backgroundColor: widget.backgroundColor,
-            elevation: widget.elevation,
-            shape: widget.shape,
-          ),
+    return AnimatedBuilder(
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/scaffold.dart]: This change doesn't seem to be related to the PR's description. If animationController is guaranteed to be non-null, are we missing some asserts? Maybe this change should be made in a separate PR?

Comment (rami-a) [reply packages/flutter/lib/src/material/scaffold.dart]: I'll undo this change as it isn't related, you're right

File: packages/flutter/lib/src/material/bottom_sheet.dart (hunk 2 above)
Comment (rami-a) [reply packages/flutter/lib/src/material/bottom_sheet.dart]: Sure thing

**Review by HansMuller — APPROVED** [review]
LGTM
